src/dbnode/persist/fs/merger_test.go

```diff
@@ -441,8 +441,11 @@ func testMergeWith(
 		persist.PreparedDataPersist{
 			Persist: func(id ident.ID, tags ident.Tags, segment ts.Segment, checksum uint32) error {
 				persisted = append(persisted, persistedData{
-					id:      id,
-					segment: segment,
+					id: id,
+					// NB(bodu): Once data is persisted the `ts.Segment` gets finalized
+					// so we can't read from it anymore or that violates the read after
+					// free invariant. So we `Clone` the segment here.
+					segment: segment.Clone(nil),
```

Comment:
@robskillington can you check that this is the right behaviour here?

I dropped the `.IncRef()` below as well because it seemed like the test code shouldn't be holding onto a ref if all it's doing is handing over all the data to the persist manager to be persisted to disk (and finalized). 

Before, it was throwing an error on Finalize while num refs > 0.

Reply:
Yeah this is fine for test code, and yeah that makes sense.